### PR TITLE
Bounty Pallet: add `approve_bounty_with_curator` call to `bounties` pallet

### PR DESCRIPTION
**Pull request by davidk-pt**
Resolves issue https://github.com/paritytech/polkadot-sdk/issues/5928

Adds `approve_bounty_with_curator` call to the `bounties` pallet to combine functions of `approve_bounty` and `propose_curator` into one call. Also adds a new status `ApprovedWithCurator` required to distinguish if bounty was approved with curator when skipping through `Funded` status and moving to `CuratorProposed` status.

If `unassign_curator` is called after `approve_bounty_with_curator` the process will fall back to the old flow of calling `propose_curator` separately.

### CODE REVIEW

File: substrate/frame/bounties/src/lib.rs
Comment (davidk-pt) [reply substrate/frame/bounties/src/lib.rs:179]: @muharem Because there still needs some time to pass when bounty is funded because as I understand only at beginning of the block bounties are funded. There could be a time when bounty is approved but funds are not available for some time. So, if we were to jump to a Funded status we would be trying to fund the bounty right away taking from the treasury, and if we don't have funds we can't fund bounty and propose a curator yet. So we have status that bounty is ApprovedWithCurator to be able to approve the bounty without having funds yet.

Correct me if I'm wrong in this

**Comment by davidk-pt** [comment]
bot help

**Comment by davidk-pt** [comment]
bot bench-all pallet --pallet=bounties

**Comment by davidk-pt** [comment]
bot bench-all pallet --pallet=bounties

**Comment by davidk-pt** [comment]
bot bench-all pallet --pallet=pallet-bounties

**Comment by davidk-pt** [comment]
bot bench-all pallet --pallet=pallet_bounties

File: prdoc/pr_5961.prdoc
Comment (muharem) [inline prdoc/pr_5961.prdoc:9]: I think we also need to mention new status variant, and how it behaves. 
Ideally we would document the status machine, right now clients has to read through the code base to understand it. It could be a doc for `BountyStatus` type.

File: substrate/frame/bounties/src/benchmarking.rs
Comment (muharem) [inline substrate/frame/bounties/src/benchmarking.rs:130]: we need to assert the event at least for this new bench to make sure your benchmarks does what is intended. you can check other pallets for examples. you can test it by running the tests with runtime-benchmarks feature enabled. no need to regenerate the weights.

File: substrate/frame/bounties/src/lib.rs
Comment (muharem) [reply substrate/frame/bounties/src/lib.rs:179]: you are right

**Review by muharem — COMMENTED** [review]
Looks good. Two minor comments

**Review by muharem — APPROVED** [review]

File: substrate/frame/bounties/src/lib.rs
Comment (kianenigma) [inline substrate/frame/bounties/src/lib.rs:829]: This call seems like is an example of #5958.

Comment (kianenigma) [reply substrate/frame/bounties/src/lib.rs:829]: But just looking at the diff, and knowing little about this pallet, I disagree with this statement. 

If this call was purely an equivalent of `batch(approve, propose)`, then:

1. The implementation should be merely a `approve().and(propose)`. Perhaps this would be less efficient, but it should be functionally the same.
2. There would be no need to introduce a new `BountyStatus` variant. 

So perhaps this PR is actually doing more than just introducing a shorthand for combining two existing atomic operations?

Comment (davidk-pt) [reply substrate/frame/bounties/src/lib.rs:829]: The issue is ( https://github.com/paritytech/polkadot-sdk/pull/5961#discussion_r1808423634 ) that after bounty is approved some time must pass until bounty reaches `Funded` status because bounty fundings happen in the beginning of every block.

The only way for this to be atomic is if with approval we try to fund the bounty right away, but there may not be enough money in the treasury and such action will revert.

I'm open to simpler solutions which may keep the flow backwards compatible with previous version.

Comment (kianenigma) [reply substrate/frame/bounties/src/lib.rs:829]: Thanks for explaining! The current solution is good, I was mainly interested in the relationship between this and #5958.

File: substrate/frame/bounties/src/lib.rs
Comment (Ank4n) [reply substrate/frame/bounties/src/lib.rs:180]: I’m not very familiar with the code, but having multiple statuses with similar meanings seems like a potential code smell.

How about modifying `Approved` with the field `maybe_curator: Option<AccountId>`? Probably we want to avoid any breaking change (which is totally understandable)? 

I leave the judgement to you guys. Though it would be great to have some doc explaining expected lifecycle of a bounty (I couldn't find one).

File: substrate/frame/bounties/src/lib.rs
Comment (Ank4n) [inline substrate/frame/bounties/src/lib.rs:869]: Do we need to emit both events? Is `BountyApproved` inferable from `CuratorProposed`?

**Review by Ank4n — APPROVED** [review]
Couple of questions but in general PR looks fine to me.

File: substrate/frame/bounties/src/lib.rs
Comment (davidk-pt) [reply substrate/frame/bounties/src/lib.rs:869]: I think so, yes, because this function does logically what `approve_bounty` and `propose_curator` does, and they emit these events separately

Comment (davidk-pt) [reply substrate/frame/bounties/src/lib.rs:869]:
```
I’m not very familiar with the code, but having multiple statuses with similar meanings seems like a potential code smell.

How about modifying Approved with the field maybe_curator: Option<AccountId>? Probably we want to avoid any breaking change (which is totally understandable)? Though any indexer now needs to support the ApprovedWithCurator status as well, so I guess breaking change in this scenario is better.

I leave the judgement to you guys. Though it would be great to have some doc explaining expected lifecycle of a bounty (I couldn't find one).
```
Wouldn't that break existing which used `Approved` enum? I think it would add at least 1 byte for the option of `Approved` whether it is `None` or `Some`, with these changes we leave the rest of the enums unmodified and people would only need to add support for the new one

File: substrate/frame/bounties/src/lib.rs
Comment (bill-pap) [reply substrate/frame/bounties/src/lib.rs:180]: Hello, this is Vasilis.
This will not be a huge deal for us on the Bounty Manager, we will just add a new status once the change is implemented, or we could also modify the current one.
As we are not running an indexer, it should in theory be a small effort for us to update it.

Comment (muharem) [reply substrate/frame/bounties/src/lib.rs:180]: the question here - should we alter existing `Approve` status by adding an optional curator or we add new `ApproveWithCurator` status. we trying to learn here from you, from clients of this API. as a code maintainer we rather alter `Approve` to keep things cleaner (but I believe the impact is low). however I think from the client perspective it is better we add new status (append new enum variation). in this case it should not be breaking change (I believe the impact is much higher).

File: substrate/frame/bounties/src/lib.rs
Comment (Ank4n) [reply substrate/frame/bounties/src/lib.rs:869]: > I think we need to emit both, because this function does logically what approve_bounty and propose_curator does, and they emit these events separately.

Only ask this since events are expensive and stored in state of each block. But of course if you have a good reason to emit both, go for it.

> Wouldn't maybe_curator: Option<AccountId> break existing which used Approved enum?

Is breaking change always bad? Especially since old clients may have unexpected side effects, i.e. just miss the approved (`ApprovedWithCurator`) status altogether if they depend on the existing `Approved` status. 

Also types are dynamically generated from substrate metadata and potentially for js clients this might not even be a breaking change.

P.S.: None of this is a huge deal tbh. I’m comfortable with either approach, as long as we’re mindful of the tradeoffs.

File: substrate/frame/bounties/src/lib.rs
Comment (bill-pap) [reply substrate/frame/bounties/src/lib.rs:180]: In my opinion, it is cleaner to alter it as well and for us it's not that much of a problem, as it will not be really a big breaking change for us because our app is pretty small and we are not indexing.

What I can imagine is that it will be a more important and/or breaking change for other platforms who might be indexing or saving this status, like e.g. Polkassembly and Subsquare. It might be actually a good idea to ask their opinions as well.

Comment (muharem) [reply substrate/frame/bounties/src/lib.rs:180]: Indexers is good example. Can you link someone from those teams?
But even for your use case, is this a breaking change? I can also imagine it to be a non-breaking, if metadata automatically updated and types decoded from it.

Comment (bill-pap) [reply substrate/frame/bounties/src/lib.rs:180]: I will ping them, as I don't know their handles.
I think even if the metadata updates and gets decoded, it will still be a "soft" break, as we will need to update some thigns to show it correctly. Most likely unless a bounty is currently in that status before we update it, I believe it won't have a breaking effect though.

Comment (wliyongfeng) [reply substrate/frame/bounties/src/lib.rs:180]: Adding new logic is better than alteration to data structure, because it cost less work to add patches.

Comment (muharem) [reply substrate/frame/bounties/src/lib.rs:180]: Thank you.
It would be helpful to create a document (or perhaps even a talk, followed by a document) to give runtime developers a clearer understanding of existing client use cases, what constitutes a breaking change for a client (less and more painful), and which changes are worthwhile. It could clarify what is generated dynamically from metadata and what isn’t, as well as whether enum variant indexes or names are used. This will likely require collaborative efforts to cover more use cases.
cc: @bill-pap @wliyongfeng @Tbaut @josepot

Comment (TarikGul) [reply substrate/frame/bounties/src/lib.rs:180]: > But my thought was - better non breaking change, all clients keep working. new feature available only to the clients adopting the new status.

Generally speaking, breaking changes should be avoided if possible (when discussing downstream clients, and tools), and if they are necessary then they should be correctly communicated in a changelog - and/or runtime release docs.

> It would be helpful to create a document (or perhaps even a talk, followed by a document) to give runtime developers a clearer understanding of existing client use cases, what constitutes a breaking change for a client (less and more painful)

Even though this would be a tall task, I totally agree. I would love to see more documentation on this front.

> It could clarify what is generated dynamically from metadata and what isn’t

Atleast from the perspective of PJS - everything is dynamically generated from the metadata now. That being said the static typing  (not to be confused with the dynamic generation at runtime) requires a few longer steps to get the types reflected in the PJS libraries - generate the static types - fix any required changes - release the api - then update all deps downstream. 

If I am not mistaken the PAPI descriptors do a way better job at supplying the types (But I am no expert - yet :))

Comment (kianenigma) [reply substrate/frame/bounties/src/lib.rs:180]: If we want to make this effort future-proof, and bring it up to understanding with #4722, we should be asking ourselves: 

`BountyStatus` is the status enum that the internal logic of the pallet would need to care about. Is it even reasonable to expose this to the users? 

For example, if we are to design a `fn status_of(bounty) -> BountyStatus`, what would the return type be?

Comment (josepot) [reply substrate/frame/bounties/src/lib.rs:180]: > How about modifying `Approved` with the field `maybe_curator: Option<AccountId>`?

I'm sorry that I'm late to the party. Modifying `Approved` to be an `Option<AccountId>` would have been better IMO.

The reason is simple: if someone is using code that doesn't take into account the new state, then if they receive the variant "`ApprovedWithCurator`" their DApp will blow up no matter what, because they don't know how to deal with that variant.

On the other hand, modifying the `Approved` variant to support an optional value, may prevent some DApps from blowing up. Also, at least with PAPI, it makes it much simpler for DApp developers to adapt to the change on their DApps.

I'm sorry that I'm sharing this after the PR has been merged.

Comment (muharem) [reply substrate/frame/bounties/src/lib.rs:180]: This means the impact can be different for the dapps and indexers. We need to learn more about it and document. It seems like we really can improve the status quo.

Comment (josepot) [reply substrate/frame/bounties/src/lib.rs:180]: I really fail to understand why this change is better for indexers.

**Pull request merged**
